cupy/_core/_routines_linalg.pyx

```diff
@@ -817,10 +814,26 @@ cpdef ndarray matmul(ndarray a, ndarray b, ndarray out=None):
                 ','.join([str(_) for _ in orig_a.shape]),
                 ','.join([str(_) for _ in orig_b.shape])))
 
+    if out is not None and out.shape != tuple(out_shape):
+        raise ValueError('Output array has an invalid size')
+
     if a.size == 0 or b.size == 0:
-        return cupy.zeros(out_shape, ret_dtype)
+        if out is None:
+            return cupy.zeros(out_shape, ret_dtype)
+        else:
+            out.fill(0)
+            return out
 
-    out = ndarray(out_shape, dtype=dtype)
+    ret = None
+    if out is None:
+        out = ndarray(out_shape, dtype=dtype)
+    elif not out.flags.c_contiguous:
+        raise NotImplementedError('Output array must be C-contiguous')
+    elif out.dtype != ret_dtype:
+        raise NotImplementedError('Output array has incorrect dtype')
+    elif dtype != ret_dtype:
+        ret = out
+        out = ndarray(out_shape, dtype=dtype)
```

Comment:
```suggestion
    if out is not None and out.dtype == dtype and out.flags.c_contiguous:
        ret = None
    else:
        if out is None and dtype != ret_dtype:
            ret = ndarray(out_shape, dtype=ret_dtype)
        else:
            ret = out
        out = ndarray(out_shape, dtype=dtype)
```

Moreover, renaming (`out`, `ret`) to (`c`, `out`) may make the code more readable, i.e.
```python
    if out is not None and out.dtype == dtype and out.flags.c_contiguous:
        c = out
    else:
        c = ndarray(out_shape, dtype=dtype)
        if out is None:
            if dtype == ret_dtype:
                out = c
            else:
                out = ndarray(out_shape, dtype=ret_dtype)
...
    if out is not c:
        elementwise_copy(c, out)
```

Reply:
OK!